### PR TITLE
lxd/init: Allow preseeding cluster_token

### PR DESCRIPTION
**Pull request by stgraber**
Signed-off-by: Stéphane Graber <stgraber@ubuntu.com>

### CODE REVIEW

File: doc/clustering.md [hunk 1]
```diff
@@ -145,6 +145,14 @@ opyQ1VRpAg2sV2C4W8irbNqeUsTeZZxhLqp4vNOXXBBrSqUCdPu1JXADV0kavg1l
     value: ""
 ```
 
+When joining a cluster using a cluster token, the following fields can be omitted:
```
Comment (tomponline) [inline doc/clustering.md:148]: Cluster join token? For consistency of terminology?

File: lxd/init.go [hunk 2]
```diff
@@ -26,6 +26,10 @@ type initDataCluster struct {
 	// The path to the cluster certificate
 	// Example: /tmp/cluster.crt
 	ClusterCertificatePath string `json:"cluster_certificate_path" yaml:"cluster_certificate_path"`
+
+	// A cluster join token
+	// Example: BASE64-TOKEN
+	ClusterToken string `json:"cluster_token" yaml:"cluster_token"`
```
Comment (tomponline) [inline lxd/init.go:32]: ClusterJoinToken for consistency and to indicate its just for joining.

File: lxd/main_init.go [hunk 3]
```diff
@@ -163,6 +166,51 @@ func (c *cmdInit) Run(cmd *cobra.Command, args []string) error {
...
+			_, _, err := net.SplitHostPort(clusterAddress)
+			if err != nil {
+				clusterAddress = fmt.Sprintf("%s:%d", clusterAddress, shared.DefaultPort)
+			}
+			config.Cluster.ClusterAddress = clusterAddress
+
+			// Cluster certificate
+			cert, err := shared.GetRemoteCertificate(fmt.Sprintf("https://%s", config.Cluster.ClusterAddress), version.UserAgent)
+			if err != nil {
+				fmt.Printf("Error connecting to existing cluster node %q: %v\n", clusterAddress, err)
```
Comment (tomponline) [inline lxd/main_init.go:192]: Cluster member father than cluster node (at least that's my understanding that we don't use node terminology in user facing statements)

**Review by tomponline — APPROVED** [review]
Minor terminology nits otherwise all good.

File: lxd/init.go (hunk 2 above)
Comment (stgraber) [reply lxd/init.go:32]: Hmm, `ClusterAddress`, `ClusterCertificate`, `ClusterPassword` are all similarly used for join despite not having their name reflect it, so this felt more in line with the rest of them.

File: lxd/main_init.go (hunk 3 above)
Comment (stgraber) [reply lxd/main_init.go:192]: Ah yeah, that's a copy/paste, I'll push a commit to tweak the rest.

File: lxd/init.go (hunk 2 above)
Comment (tomponline) [reply lxd/init.go:32]: My thinking was that those don't expire after joining though, and the comments all say join token but field name was slightly different. I see what you mean though.

Comment (stgraber) [reply lxd/init.go:32]: `ClusterAddress` kinda expires after joining as that gets updated with every hearbeat. `ClusterCertificate` expires the next time you use `lxc cluster update-certificate`, `ClusterPassword` as used in the past would expire almost immediately (admin unsetting it).

Comment (tomponline) [reply lxd/init.go:32]: Yeah im not dead set against it. I was mainly thinking from an integration point of view as our docs talk about a join token (and not join password for instance) and the field doesn't match that term, so might think its something else. I think most folks would take the join context into account and/or figure it out from the api docs themselves. I just thought I'd flag it as it triggered my inconsistency alarm 😀 happy either way.

Comment (stgraber) [reply lxd/init.go:32]: yeah, I'm not super keen on making the key even longer than it is. I'm actually kinda confused as to why all those keys have `cluster_` as a prefix when they're all already part of a cluster API... but not a lot we can do about that now.

**Comment by stgraber** [comment]
@tomponline should be good to go

**Pull request merged**